test/storage/bplustree_test.cpp

```diff
@@ -0,0 +1,1468 @@
+#include <cstdlib>
+#include <set>
+#include <unordered_map>
+
+#include "storage/index/bplustree.h"
+#include "storage/storage_defs.h"
+#include "test_util/multithread_test_util.h"
+#include "test_util/test_harness.h"
+
+namespace noisepage::storage::index {
+
+unsigned int globalseed = 9;
+
+class BPlusTreeTests : public TerrierTest {
+ public:
+  const uint32_t num_threads_ = 4;
+  common::WorkerPool thread_pool_{num_threads_, {}};
+
+ protected:
+  void SetUp() override { thread_pool_.Startup(); }
+
+  void TearDown() override { thread_pool_.Shutdown(); }
+};
+
+void BasicNodeInitializationInsertReadAndFreeTest() {
+  auto bplustree = new BPlusTree<int, TupleSlot>;
+  BPlusTree<int, TupleSlot>::KeyNodePointerPair p1;
+  BPlusTree<int, TupleSlot>::KeyNodePointerPair p2;
+
+  // Get inner Node
+  auto node = BPlusTree<int, TupleSlot>::ElasticNode<BPlusTree<int, TupleSlot>::KeyNodePointerPair>::Get(
+      10, BPlusTree<int, TupleSlot>::NodeType::LeafType, 0, 10, p1, p2);
+
+  // To check if we can read what we inserted
+  std::vector<BPlusTree<int, TupleSlot>::KeyNodePointerPair> values;
+  for (unsigned i = 0; i < 10; i++) {
+    BPlusTree<int, TupleSlot>::KeyNodePointerPair p1;
+    p1.first = i;
+    values.push_back(p1);
+    node->PushBack(p1);
+    EXPECT_EQ(node->GetSize(), i + 1);
+  }
+
+  using ElementType = BPlusTree<int, TupleSlot>::KeyNodePointerPair;
+
+  unsigned i = 0;
+  for (ElementType *element_p = node->Begin(); element_p != node->End(); element_p++) {
+    EXPECT_EQ(element_p->first, i);
+    i++;
+  }
+
+  // To Check if we are inserting at the correct place
+  EXPECT_EQ(reinterpret_cast<char *>(node) + sizeof(BPlusTree<int, TupleSlot>::ElasticNode<BPlusTree<int, TupleSlot>>),
+            reinterpret_cast<char *>(node->Begin()));
+
+  EXPECT_EQ(&(node->GetLowKeyPair()), node->GetElasticLowKeyPair());
+  EXPECT_EQ(&(node->GetHighKeyPair()), node->GetElasticHighKeyPair());
+  EXPECT_EQ(node->GetLowKeyPair().first, p1.first);
+  EXPECT_EQ(node->GetHighKeyPair().first, p2.first);
+  EXPECT_NE(&p1, &(node->GetLowKeyPair()));
+  EXPECT_NE(&p2, &(node->GetHighKeyPair()));
+
+  // Free the node - should not result in an ASAN
+  node->FreeElasticNode();
+  delete bplustree;
+}
+
+void InsertElementInNodeTest() {
+  auto bplustree = new BPlusTree<int, TupleSlot>;
+  BPlusTree<int, TupleSlot>::KeyNodePointerPair p1;
+  BPlusTree<int, TupleSlot>::KeyNodePointerPair p2;
+
+  // Get inner Node
+  auto node = BPlusTree<int, TupleSlot>::ElasticNode<BPlusTree<int, TupleSlot>::KeyNodePointerPair>::Get(
+      10, BPlusTree<int, TupleSlot>::NodeType::LeafType, 0, 10, p1, p2);
+
+  for (unsigned i = 0; i < 10; i++) {
```

Comment:
Similar comment to the one above. I think it would be better to name "i" to something more meaningful.

Reply:
Done